packages/browser/scripts/generate-rules.js

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env node
+
+const { builtinModules } = require('repl');
+const { promises: fs } = require('fs');
+const { resolve } = require('path');
+const globals = require('globals');
+
+/**
+ * The path to write the rules to.
+ *
+ * @type {string}
+ */
+const RULES_PATH = resolve(__dirname, '../src/rules.json');
+
+/**
+ * A list of messages to use for specific Node.js built-in modules. This will
+ * show up in the error message when a restricted import is used.
+ *
+ * @type {Record<string, string>}
+ */
+const MESSAGE_OVERRIDES = {
+  buffer: `Use 'Uint8Array' instead.`,
+  Buffer: `Use 'Uint8Array' instead.`,
+  crypto: `Use 'ethereum-cryptography' instead.`,
+};
+
+/**
+ * Get a rule for a given module name.
+ *
+ * @param {string} name - The module name.
+ * @returns {{ name: string; message: string; }[]} An array of
+ * restricted modules.
+ */
+const getModuleRule = (name) => {
+  const message =
+    MESSAGE_OVERRIDES[name] || 'This module is not available in the browser.';
+
+  return [
+    {
+      name,
+      message,
+    },
+    {
+      name: `node:${name}`,
+      message,
+    },
+  ];
+};
+
+/**
+ * Get a list of restricted globals. This is a list of all globals that are
+ * not available in both the browser and Node.js.
+ *
+ * @returns {{ name: string; message: string; }[]} An array of
+ * restricted globals.
+ */
+const getGlobals = () => {
+  const items = Array.from(
+    new Set([...Object.keys(globals.browser), ...Object.keys(globals.node)]),
+  );
+
+  // Find all globals that are not in the `shared-node-browser` list, i.e.,
+  // globals that are only in the `browser` or `node` list.
+  return items
+    .filter(
+      (item) => !Object.keys(globals['shared-node-browser']).includes(item),
+    )
+    .map((name) => ({
+      name,
+      message:
+        MESSAGE_OVERRIDES[name] ||
+        'This global is not available in both the browser and Node.js.',
+    }));
+};
+
+const generateRules = () => {
+  return {
+    // Disallow use of Node.js built-in modules.
+    'no-restricted-imports': [
+      'error',
+      {
+        paths: builtinModules.map(getModuleRule).flat(),
+      },
+    ],
+    'no-restricted-globals': ['error', ...getGlobals()],
```

Comment:
If we agree on adding `no-undef` to our TypeScript config, then I don't think we need this rule anymore. The ESLint environment should be enough to explain what globals exist in that environment; we only need this rule if we want to include a global in the environment but still prevent it from being used.

Reply:
Yeah, that seems like a better solution, and would also remove the need for rule generation combined with the other proposed rule.

Reply:
Apparently `no-undef` also breaks stuff that is available in both Node.js and browsers.

```
'console' is not defined.
```

Reply:
Hmm. That should be brought in by the `shared-node-browser` environment. Maybe the base config environment is getting overwritten in the typescript config? If so we could re-add it.

Reply:
Oh, right. I just had to add an environment to the `.eslintrc.js` we use to lint this package itself. 😄

Reply:
Where were you seeing this error? Locally I am not seeing it, even after removing the environment.

I thought using `extend` would bring the environments along, so they wouldn't need to be re-declared.

Reply:
Removed here: https://github.com/MetaMask/eslint-config/pull/255

Reply:
Might have been just the test that needs it. The test doesn't use any config other than the one being tested.